File: src/main/java/org/hisp/dhis/expression/ast/DataItemModifier.java
package org.hisp.dhis.expression.ast;

/**
 * A.K.A "dot functions"
 *
 * @author Jan Bernitt
 */
@SuppressWarnings("java:S115")
public enum DataItemModifier implements Typed
{
    aggregationType(ValueType.STRING),
    maxDate(ValueType.DATE),
    minDate( ValueType.DATE),
    periodOffset(ValueType.NUMBER),
    stageOffset(ValueType.NUMBER);

    private final ValueType parameterType;

    DataItemModifier(ValueType parameterType) {
        this.parameterType = parameterType;
    }

    @Override
    public ValueType getValueType() {
        return ValueType.SAME;
    }

    public ValueType getParameterType() {
        return parameterType;
    }
}

File: src/jvmMain/java/org/hisp/dhis/lib/expression/ast/DataItemModifier.kt
package org.hisp.dhis.lib.expression.ast

import org.hisp.dhis.lib.expression.spi.ValueType

/**
 * A.K.A "dot functions"
 *
 * @author Jan Bernitt
 */
enum class DataItemModifier(vararg parameterTypes: ValueType) : Typed {

    aggregationType(ValueType.STRING),
    maxDate(ValueType.DATE),
    minDate(ValueType.DATE),
    periodOffset(ValueType.NUMBER),
    stageOffset(ValueType.NUMBER),
    yearToDate,
    periodAggregation,
    subExpression;

    @JvmField
    val parameterTypes: List<ValueType>

    init {
        this.parameterTypes = listOf(*parameterTypes)
    }

    override fun getValueType(): ValueType {
        return ValueType.SAME
    }
}
